### ISSUE TITLE
Fix S2178 FP/FN: Incorrect suggestion when using the LINQ Aggregate extension

### ISSUE THREAD

**Issue by siewers** [SonarSource/sonar-dotnet#8527]
### Description

I want to return `true` if any is `true` otherwise `false`.

### Repro steps

```cs
var validItems = new List<string>();
var items = new[] { "this_is_long_enough", "bar", "baz", "this_is_also_long_enough" };
var hasValidItems = items.Aggregate(false, (current, item) => current | IsValidItem(item));
//                                                                    ^
//                                               SonarLint: Use short-circuiting operators

// validItems contains "this_is_long_enough" and "this_is_also_long_enough"

bool IsValidItem(string item)
{
    if (item.Length > 5)
    {
        validItems.Add(item);
        return true;
    }

    return false;
}
```

### Expected behavior

The warning is a false positive when I need to evaluate all items.

### Actual behavior

I'm told to use short-circuiting operators ( `||`) , which results in a different behavior.
If I modify the code to adhere to the rule, it results in the above code example not adding the last valid item from the list.
```cs
// validItems only contains "this_is_long_enough"
```

### Known workarounds

Ignore the warning or rewrite the code to use a compound assignment instead:
```cs
var items = new[] { "foo", "bar", "baz", "this_is_long_enough" };
var hasValidItems = false;

foreach (var item in items)
{
    hasValidItems |= IsValidItem(item);
}

bool IsValidItem(string item)
{
    return item.Length > 5;
}
```

### Related information
C# 11.0
.NET 7.0
JetBrains Rider 2023.3.2
SonarLint 10.2.1.77304

**Comment by Tim-Pohlmann** [comment]
Hi @siewers, thanks for the report!
The problem stems from having a function with side-effects (adding an item to the list) in a boolean condition. This usually leads to code that can easily be misunderstood. In other words, if following S2178's suggestion leads to different results, we recommend rewriting your code to be more straightforward. Thus, I consider this a true positive.

**Comment by siewers** [comment]
Thanks, but I don't understand why the `for-each` doesn't trigger the same warning?

**Comment by Tim-Pohlmann** [comment]
```
foreach (var item in items)
{
    hasValidItems |= IsValidItem(item);
}
```
Does not raise, because there is no `||=` operator. You could rewrite it as `hasValidItems = hasValidItems || IsValidItem(item);` but that would be a much bigger change.

**Comment by siewers** [comment]
Okay, I'm not sure I understand the argument of it being a much bigger change. 
If it's violating the rule, the warning should be present, regardless of a code-fix being available.
What confuses me is that I can rewrite the code to do the same thing, but prevent the warning from appearing.

**Comment by siewers** [comment]
I would also claim that the code fix is wrong because it alters the behavior of the code, without considering the context. 
Sure, it might be a valid suggestion because the code _might_ be harder to understand, but providing a fix that fundamentally alters the behavior is in my opinion not correct.

**Comment by Tim-Pohlmann** [comment]
The rule is about using short-circuit logic where possible. The main goal is to avoid issues like this:
```
return x is null | x.IsEmpty();   // throws NullReferenceException because both sides are evaluated
return x is null || x.IsEmpty();  // null safe
```
When using the compound assignment `|=`, this is usually not a problem.

Regarding the code fix: The code fix does what it says, so on its own, it is okay. You are right, however, that, combined with the rule description, this can lead the developer into a trap. I looked deeper into this and found an optional longer description in our internal documents, but the .NET analyzer does not implement it. I think it would significantly improve the situation and created an issue to add it: #8559

Thanks again for the report and extended explanations.
